compiler/compiler/src/run.rs

```diff
@@ -70,43 +76,176 @@ pub struct Case {
+            Self::Accepted => "accepted".fmt(f),
+            Self::Rejected => "rejected".fmt(f),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum EvaluationStatus {
+    Success,
+    Failed(String),
+}
+
+impl fmt::Display for EvaluationStatus {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Success => write!(f, "success"),
+            Self::Failed(e) => write!(f, "failed: {e}"),
         }
     }
 }
 
 /// All details about the result of a case that was run.
 #[derive(Debug, Clone)]
-pub struct CaseOutcome {
-    pub program_name: String,
-    pub function: String,
-    pub status: Status,
-    pub verified: bool,
-    pub execution: String,
-    pub output: Value,
+pub enum Outcome {
+    ExecutionOutcome {
+        program_name: String,
+        function: String,
+        status: ExecutionStatus,
+        verified: bool,
+        execution: String,
+        output: Value,
+    },
+    EvaluationOutcome {
+        program_name: String,
+        function: String,
+        status: EvaluationStatus,
+        output: Value,
+    },
+}
+
+impl Outcome {
+    pub fn output(&self) -> Value {
+        match self {
+            Self::ExecutionOutcome { output, .. } => output.clone(),
+            Self::EvaluationOutcome { output, .. } => output.clone(),
+        }
+    }
+}
+
+pub fn run_without_ledger(config: &Config, cases: &[Case]) -> Result<Vec<Outcome>> {
+    if cases.is_empty() {
+        return Ok(Vec::new());
+    }
+
+    let programs = config
+        .programs
+        .iter()
+        .map(|Program { bytecode, .. }| {
+            // Parse the program.
+            let program = snarkvm::prelude::Program::<CurrentNetwork>::from_str(bytecode).unwrap();
+            // Return the program and its name.
+            Ok(program)
+        })
+        .collect::<Result<Vec<_>>>()?;
+
+    let programs_and_editions = programs
+        .into_iter()
+        .map(|program| {
+            let edition = 1;
+            (program, edition)
+        })
+        .collect::<Vec<_>>();
+
+    let rng = &mut ChaCha20Rng::seed_from_u64(config.seed);
+
+    let mut outcomes = Vec::new();
+
+    for case in cases {
+        let bad_outcome = |e: String| Outcome::EvaluationOutcome {
+            program_name: case.program_name.clone(),
+            function: case.function.clone(),
+            status: EvaluationStatus::Failed(e),
+            output: Value::make_unit(),
+        };
+
+        let vm = VM::from(ConsensusStore::<CurrentNetwork, ConsensusMemory<CurrentNetwork>>::open(
+            StorageMode::Production,
+        )?)?;
+
+        vm.process().write().add_programs_with_editions(&programs_and_editions)?;
+
+        let private_key = PrivateKey::from_str(leo_ast::TEST_PRIVATE_KEY).unwrap();
+        let program_id = ProgramID::<CurrentNetwork>::from_str(&case.program_name).unwrap();
+        let function_id = Identifier::<CurrentNetwork>::from_str(&case.function).unwrap();
```

Comment:
For these `cases`, do we `unwrap` because the program name and function are validated eariler?

Reply:
No longer unwrapping. Might as well make them into bad outcomes.